File: typescript-kata/src/algorithms/searching/binary-search.ts
export function BinarySearch(haystack: number[], needle: number): boolean {
  let lo: number = 0;
  let hi: number = haystack.length;

  while (lo < hi) {
    const mid = Math.floor(lo + (hi - lo) / 2);
    const val = haystack[mid];

    if (val === needle) {
      return true;
    } else if (val > needle) {
      hi = mid;
    } else {
      lo = mid + 1;
    }
  }

  return false;
}

File: python_kata/src/algorithms/searching/binary_search.py
def binary_search(haystack: list[int], needle: int) ->bool:
    
    lo = 0
    hi = len(haystack)

    while lo < hi:
        mid = lo + (hi - lo) // 2
        val = haystack[mid]

        if val == needle:
            return True
        elif val > needle:
            hi = mid
        else:
            lo = mid + 1

    return False
